azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py

```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
             pass  # is there any other case
 
     def _on_provider_unsubscribe_complete(self, mid):
-        """
-        Callback that is called by the provider when a subscribe operation is complete.
-        """
-        logger.info("_on_provider_unsubscribe_complete")
-        callback = self._unsubscribe_callback
-        if callback:
-            self._unsubscribe_callback = None
+        if mid in self._callback_map:
+            callback = self._callback_map[mid]
+            del self._callback_map[mid]
             callback()
+        else:
+            # TODO: tests for unkonwn MID cases
+            logger.warning("UNSUBACK received with unknown MID: %s", str(mid))
 
-    def _add_event_to_queue(self, event_data):
+    def _add_action_to_queue(self, event_data):
         """
-        Queue an event for sending later.  All events that get sent end up going into
-        this queue, even if they're going to be sent immediately.
+        Queue an action for running later.  All actions that need to run while connected end up in
+        this queue, even if they're going to be run immediately.
         """
-        logger.info("Adding event to queue for later sending")
-        # TODO: throw here if event_data.args[0] is not an event/message
-        self._event_queue.put_nowait((event_data.args[0], event_data.args[1]))
+        action = event_data.args[0]
+        if isinstance(action, TransportAction):
+            self._action_queue.put_nowait(event_data.args[0])
+        else:
+            assert False
+
+    def _do_single_action(self, action):
+        if isinstance(action, SendMessageAction):
+            logger.info("running SendMessageAction")
+            message_to_send = action.message
+            base_topic = self._get_telemetry_topic()
+
+            if isinstance(message_to_send, Message):
+                encoded_topic = _encode_properties(message_to_send, base_topic)
+            else:
+                encoded_topic = base_topic
+                message_to_send = Message(message_to_send)
```

Comment:
Is this a thing we want to do in the transport, or in the client?

I've already implemented this Message wrapping in the client, but I could ditch it in favor of doing it here.

Alternatively, we could remove it here, and keep it in the client. What do you think the most appropriate place to do this is?

Reply:
I think the client is the right place.  It has to happen for all transports.  I removed it from here.

---
In reply to: [266169151](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266169151) [](ancestors = 266169151)